Share consensus-critical constants through parameter

linked0 opened the pull request:
For unit tests, we need to modify the protocol-level constants because running the network based on the defined constants like`ValidatorCycle`, or 1008 make unit tests very time-consuming and error-prone, so we need to make the contents variable in unit tests. 

I applied the concept of passing parameters to the code, so we can set the constants when calling the `runNode` function for the real environment or calling the `makeTestNetwork` function for test environments.

Relates to #894

linked0 committed:
Add `ConsensusParams` class for consensus-critical constants

For unit-tests, it's not efficient to test with the constants for consensus
like the `ValidatorCycle` because the original value make the tests slower and
very time-consuming. The goal of unit tests is checking the operations run
properly in test environment.

In this commit, we provide the definition for configurable consensus-critical
constants, or the `ConsensusParams`.

linked0 committed:
Propagate ConsensusParams through the code

In this commit, we start the network with the set constants in `runNode`
function for real environments or `makeTestNetwork` for the test
environments.

linked0 committed:
Switch ValidatorCycle from constant to ConsensusParams

The constants is already contained in the `ConsensusParams`, so the
`Enrollment` does not need to have the constants. And we replace the
`Enrollment.ValidatorCycle` with the `ValidatorCycle` variable of the
`ConsensusParams` object.

linked0 commented on source/agora/consensus/data/Params.d:
```diff
@@ -0,0 +1,66 @@
+
+    ***************************************************************************/
+
+    public this (uint validator_cycle)
```
```suggestion
    public this (uint validator_cycle = 1008)
```

linked0 commented on source/agora/consensus/data/Params.d:
```diff
@@ -0,0 +1,66 @@
+class Params
+{
+    /// The cycle length for a validator
+    public immutable uint ValidatorCycle;
```
```suggestion
    public immutable uint ValidatorCycle = 1008;
```

linked0 commented on source/agora/consensus/data/Params.d:
```diff
@@ -0,0 +1,66 @@
+
+module agora.consensus.data.Params;
+
+class Params
```
```suggestion
public immutable class Params
```

linked0 commented on source/agora/node/Validator.d:
```diff
@@ -52,10 +53,10 @@ public class Validator : FullNode, API
 
     /// Ctor
-    public this (const Config config)
+    public this (const Config config, const Params params = null)
```
```suggestion
    public this (const Config config, const ConsensusParams cparams)
```

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -821,14 +826,15 @@ public struct TestConf
 
 public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager)(
-    in TestConf test_conf)
+    in TestConf test_conf, in Params params = null)
```
`TestConf` is already supposed to carry all test-related configuration. So I suggest moving `Params` inside `TestConf`.

linked0 replied:
> `TestConf` is already supposed to carry all test-related configuration. So I suggest moving `Params` inside `TestConf`.

I'll check your suggestion. Thanks.

AndrejMitrovic commented:
If you make it a `struct`, then you won't have to do `cast(immutable)` when passing it across threads, and you won't have to do `assert(params !is null)` if you pass the struct by `const ref`.

That's just one idea though.

linked0 commented:
> If you make it a `struct`, then you won't have to do `cast(immutable)` when passing it across threads, and you won't have to do `assert(params !is null)` if you pass the struct by `const ref`.

The original idea is to have only one object over a node. But as you suggested, `struct` could be used for that with `const ref`. I will think about it.

linked0 committed:
Modify the test code for various validator cycles

In test code for the enrollment process, we make the test set the validator
cycle.

linked0 commented:
After a self-review, I will request reviews.

AndrejMitrovic commented:
If you make it a struct it becomes easy to construct:

```
struct ConsensusParams
{
    const uint cycle_length = 1008;
    immutable Block gen_block = UnitTestGenesisBlock;
}

ConsensusParams params = { cycle_length : 10 }  // just override cycle
ConsensusParams params = { gen_block : makeGenBlock(...) }  // just override genesis block
```

This is better than `auto params = new ConsensusParams(10, 300, 12345)` because it makes it easy to determine what we're initializing, and it also makes it possible to leave all other parameters be the default except the ones we want to manipulate in the tests (like shown in the example above).

This also makes it easy to pass this type to the Node's constructor even if it's running in a different thread. 

And also we should never cast objects to immutable.

AndrejMitrovic commented:
Also you could make it a default parameter, leading to less code changed. `ConsensusParams cparams = ConsensusParams.init` for example.

AndrejMitrovic commented:
And it's ok to have it as a default parameter like this because the defaults will be used in production for CoinNet.

AndrejMitrovic commented:
@Geod24 I'd like to hear your opinion here too before anything is changed.

linked0 commented:
@AndrejMitrovic Thank you for your suggestions. I'll modify as you suggested after getting @Geod24 's opinion.

> This also makes it easy to pass this type to the Node's constructor even if it's running in a different thread.

I'm not sure I follow you. I know there can be several Nodes separately in each thread and `ConsensusParams` objects exist separately in each thread. Can you explain why your suggested code makes the passing the parameter easy even if a node is running in a different thread?

AndrejMitrovic commented:
> I'm not sure I follow you. I know there can be several Nodes separately in each thread and ConsensusParams objects exist separately in each thread.

No they don't. You are casting the object to immutable, and passing a pointer to it to all other threads.

AndrejMitrovic commented:
You could do `new immutable(ConsensusParams)(...)` and that would be safe.

Geod24 commented:
Or use an `immutable class`.

Geod24 commented:
I think either way (class or struct) makes sense, depending on how you approach the problem.
However I think what is important is to have fields (or the `class` instance if we use a `class`) be at least `const` (needs to be `immutable` for the `class`).

Geod24 commented on source/agora/consensus/data/ConsensusParams.d:
```diff
@@ -0,0 +1,62 @@
+
+    Parameter set for consensus-critcal constants
+
+    FIXME: Modfiy the description
```
👋

Geod24 commented on source/agora/consensus/data/ConsensusParams.d:
```diff
@@ -0,0 +1,62 @@
+/*******************************************************************************
+
+    Parameter set for consensus-critcal constants
```
```suggestion
    Parameter set for consensus-critical constants
```

Geod24 commented on source/agora/consensus/data/ConsensusParams.d:
```diff
@@ -0,0 +1,62 @@
+    public this (uint validator_cycle)
+    {
+        this.ValidatorCycle = validator_cycle;
+    }
```
As discussed, if we use a class, we can achieve `struct`-like default constructor by doing:
```suggestion
    /***************************************************************************

        Constructor

        Params:
            validator_cycle = cycle length for a validator

    ***************************************************************************/

    public this (uint validator_cycle = 1008)
    {
        this.ValidatorCycle = validator_cycle;
    }
```

AndrejMitrovic replied:
> struct-like default constructor

I think the class is only going to become awkward to use as we add more parameters to it.

For example let's say we have:

```
this (uint validator_cycle = 1008, uint tx_in_blocks = 8, uint min_validators = 3)
```

What if I want to instantiate `ConsensusParams` and only change `min_validators` and leave the rest to their defaults? I would have to write code like this:

```D
const def = new ConsensusParams();
auto actual_cparams = new immutable ConsensusParams(def.ValidatorCycle, def.TxInBlocks, 10);
```

If it was a struct all I would have to do is:

```D
ConsensusParams params = { min_validators : 10 }
```

It's more readable and composable.

linked0 commented:
Applying class for the `ConsensusParams` would be better considering the concept of the consensus-critical constants because we intend to have only one object and share that in the network although we can do that with `struct`.

I will apply `immutable` for the class.

linked0 commented:
Ready for review.

AndrejMitrovic commented on source/agora/test/NetworkManager.d:
```diff
@@ -98,9 +100,10 @@ unittest
         ///
-        public this (immutable(Block)[] blocks)
+        public this (immutable(Block)[] blocks,
+            immutable(ConsensusParams) params = null)
```
```suggestion
            immutable(ConsensusParams) params)
```

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -986,11 +993,14 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager)
         break;
     }
 
+    immutable const_params =
+        (params !is null) ? params : new immutable(ConsensusParams)();
```
Technically this is not const anymore. Maybe:

```suggestion
    immutable cons_params =
        (params !is null) ? params : new immutable(ConsensusParams)();
```

So it refers to consensus, rather than "constant".

linked0 replied:
Right, it should have the meaning of consensus.

AndrejMitrovic commented on source/agora/consensus/data/ConsensusParams.d:
```diff
@@ -0,0 +1,49 @@
+
+module agora.consensus.data.ConsensusParams;
+
+public immutable class ConsensusParams
```
```suggestion
/// Ditto
public immutable class ConsensusParams
```

So it appears in the documentation.

linked0 commented:
I applied all the suggestions, so ready for review.

AndrejMitrovic commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -41,19 +42,24 @@ public class EnrollmentPool
     private ManagedDatabase db;
 
+    /// Parameters for consensus-critical constants
+    private immutable(ConsensusParams) params;
```
I think it's not needed in this class. I can't think of any consensus parameters that would affect how the pool behaves.

linked0 replied:
The consensus-critical constants are protocol-level values, so it must affect all the objects in the network. But the pool does not use this value now. so it can be deleted. I will delete it.

AndrejMitrovic commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -320,28 +327,31 @@ unittest
     seed_sources[utxo_hash] = Scalar.random();
-    auto enroll = createEnrollment(utxo_hash, key_pair, seed_sources[utxo_hash]);
+    auto enroll = createEnrollment(utxo_hash, key_pair, seed_sources[utxo_hash],
+        pool.params.ValidatorCycle);
```
Yeah this is unrelated to the pool. The longevity of a validator's cycle is unrelated to how long an enrollment stays in the pool.

linked0 replied:
I can change it with a `ConsensusParams' object. Thanks.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -120,8 +127,8 @@ public class Ledger
         // we are only interested in the last 1008 blocks,
```
```suggestion
        // we are only interested in the last `ValidatorCycle` blocks,
```

AndrejMitrovic replied:
Actually there are many places in the documentation which still reference `1008` as a number. We might want to fix this in another PR though, don't worry.

linked0 replied:
I'll make a PR for it later.

Geod24 commented on source/agora/consensus/data/ConsensusParams.d:
```diff
@@ -0,0 +1,50 @@
+    The set for consensus-critical constants
+
+    This defines the class for the consensus-critical constants. Only one
+    object is being created and being shared in the network. It can be very
+    useful to set the constanst from the start of the network. The
+    consconsensus-critcal constants are the protocol-level contants, so it
+    is only modified for test environments.
```
```suggestion
    object should exists for a single node. The `class` is `immutable`, hence
    the constants needs to be set at the start of the process. The
    consensus-critical constants are the protocol-level constants, so they
    shouldn't be modified outside of for test environments.
```

Please re-read documentation. Any simple spell-checker would catch typos such as "constanst", "consconsensus", "critcal" and "contents". A couple typos here and there happens, but that's a lot.

linked0 replied:
> Please re-read documentation. Any simple spell-checker would catch typos such as "constanst", "consconsensus", "critcal" and "contents". A couple typos here and there happens, but that's a lot.

Sorry for the many typos. You had already suggested a plug-in for VS Code but it doesn't work. I will find other plug-in or the way I don't make typos. Thanks.

Geod24 reviewed (changes requested):
Nice job. Just a few nitpicks and we're good, + what @AndrejMitrovic said. Additionally regarding the commit, the second should be named "Propagate ConsensusParams through the code" because that's what it does (it doesn't actually *use* it) and the third one could be named "Switch ValidatorCycle from constant to ConsensusParams".

Geod24 commented:
Note: I had some outdated comments from a previously uncompleted review. Ignore those.

linked0 commented:
Ready for review.

Geod24 reviewed (approved):
LGTM. @AndrejMitrovic do you want to have a last look ? Otherwise ready for merging.
